src/masternodes/masternodes.cpp

```diff
@@ -920,13 +920,25 @@ CAmount CCollateralLoans::precisionRatio() const
     return ratio > maxRatio / precision ? -COIN : CAmount(ratio * precision);
 }
 
-ResVal<CAmount> CCustomCSView::GetAmountInCurrency(CAmount amount, CTokenCurrencyPair priceFeedId, bool useNextPrice, bool requireLivePrice)
+ResVal<CAmount> CCustomCSView::GetAmountInCurrency(CAmount amount, CTokenCurrencyPair priceFeedId, bool useNextPrice, bool requireLivePrice, bool reverseDirection)
 {
         auto priceResult = GetValidatedIntervalPrice(priceFeedId, useNextPrice, requireLivePrice);
         if (!priceResult)
             return std::move(priceResult);
 
         auto price = priceResult.val.get();
+
+        if (reverseDirection)
+        {
+            if (price == 0)
+                return Res::Err("Price is zero (%s - %s/%s)", GetDecimaleString(price), priceFeedId.first, priceFeedId.second);
+            auto amountInCurrency = DivideAmounts(amount, price);
+            if (MultiplyAmounts(amountInCurrency, price) != amount)
+                amountInCurrency += 1;
```

Comment:
The reverse direction is only needed when overpaying.

Reply:
Both `MultiplyAmounts` and `DivideAmounts` round down. So I'd say adding this 1 here is okay.

Reply:
AFAIU DivideAmounts doesn't round up.

sample case:
price= 1000
amount = 1 fi
amountInCurrency = 0 -> increased to 1
so we return an amountInCurrency which is 1000x in value of the given amount. This magnitude of "rounding Error" shouldn't be done "silently" within this function IMHO. (on the long run this would lead to a similar error like the 1 fi swap-bug on the DEX)

Reply:
> The reverse direction is needed when overpaying.

good point. in that case I would suggest to only use it in case of overpaying, and not everytime then.
AFAIU your proposed changes in mn_checks already does that 👍

Reply:
Just to break the problem little bit and explain some aspects:

- reverse direction is necessary as we don't have oracle feeds on price of USD/DFI. So when user inputs how much DFI he wants to pay we need to convert it to DUSD, then sub the interest and see how much is left for subtracting the base of loan, and in the end those amounts mut be converted again to DFI to sub the balance and here we need reverse direction. What could be done to make this easier is suggested in below comment and I'm thinking of it now is that we could burn the complete DFI in one go and just go one direction to see how much of that is in DUSD and sub the interest and loan part.

- The part with celing 1 satoshi on DFI side is not a bad thing, it basically insures that in the price conversion you don't get to cover some satoshis less on the loan because of the conversions in price, but in the end if the covering of loan costs 1 (or few) more satoshi on the DFI side, that is just the price of it for using DFI to cover the DUSD loan. But 1 satoshi is better than dex price ;)

- I will try to make this simpler, and I ilike the solution for just to sub the amount of DFI inputed and then calc what is that in DUSD and sub interest and what is left on the base.

Reply:
The only problem I see with easier solution is if you overpay. Then we need to convert back from DUSD to DFI to see what is the exact amount of DFI we need to burn.

Reply:
> * The part with celing 1 satoshi on DFI side is not a bad thing, it basically insures that in the price conversion you don't get to cover some satoshis less on the loan because of the conversions in price, but in the end if the covering of loan costs 1 (or few) more satoshi on the DFI side, that is just the price of it for using DFI to cover the DUSD loan. But 1 satoshi is better than dex price ;)

agreed for the usecase of paying back a loan with DFI. But noone guarantees that this method will only ever be used for that. Having this behaviour in here "implicitly", might lead to problems down the road when someone uses it without realizing this implications. thats my only concern.

Reply:
Currently the feature is hardcoded to payback the DUSD loan when you supply DFI

Reply:
> Currently the feature is hardcoded to payback the DUSD loan when you supply DFI

exactly. "currently". But this method is only part of the change, not part of the feature. it might and will definitly be used in other contexts further down the road. Thats why its not called "getDUSDPerDFIPrice" right?
please always consider methods as standalone contract, not only in the way they are used right now.

Reply:
For v3 it will probably be refactored to be able to payback any dToken in loan. But right now, for the next hardfork it was decided to imolement this way and limit for DUSD only.

Reply:
ok. just wanted to raise awareness.

Reply:
I have refactored the logic of payback with DFI to be little bit simpler.
- Now reverse direction will be done only in case of overpay, because it needs to be calculated how much of DFI should be really burnt and the rest not touched.
NOTE: overpay will be common way to payback because you are not sure in which block your tx is going to end up so if it doesn't get in next block but after the next then interest will be higher, so overpay assures you have cleared whole loan if that is the intention but only necessary amount of DFI will be used.
- All other loan paybacks that are not overpay with DFI will only be calced in direction of DFI -> DUSD and then loan and interest will be subtracted from loan account and DFI will be burnt in whole amount (not interest and loan separate), so there is no need for reverse direction.

Still waiting for #998 to add DFI payback fee.

Reply:
Thanks @kuegi for the reviews and the observations. Just sharing some more context around DUSD hard coding: Given the short time for the hard fork - we didn't want to change the protocol message structure to allow all tokens than just DFI for payback - since that requires changes to all downstream APIs and SDKs. So we decided to do this in 2 phases:

Phase 1: Just do the minimal change set for DUSD.
Phase 2: For v3, this will go through another refactor making it generic along with the changes to the payback structure.